keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+  condition.
+- Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of
+  node maintenance. Then the node maintenance would be garbage collected and the node made
+  schedulable again.
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from a given node. This should complement the node's `.spec.unschedulable`,
+which prevents new pods from being scheduled on such a node. We will focus primarily on
+`kubectl drain`, but the API can also be used by other drain implementations as well (e.g. node
+autoscalers).
+
+We will first introduce the NodeMaintenance API and then later modify the behavior of the
+Kubernetes system to fix all the node drain issues we mentioned earlier. The name, NodeMaintenance
+is subject to change; please see the alternatives section for other suggestions.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration with a condition. The pods would be selected according to
+the node at first, but the selection mechanism can be extended later. Controllers can then implement
+the migration. The advantage of this approach is that controllers do not have to be aware of the
+NodeMaintenance object (no RBAC changes required). They only have to observe pods they own and react
+by migrating them. The first candidate is a deployment controller, since its workloads support
+surging to another nodes. This would help to eliminate downtime not only for single replica
+applications, but for HA applications as well. 
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I also want to observe the node drain via the API and check on
+its progress. I also want to be able to discover workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
```

Comment:
While it could be out of scope of this kep: With this change I still not see a good way of determening the drain status.
How does an admin know if all relevant pods were evicted - or not?

Reply:
please see https://github.com/kubernetes/enhancements/pull/4213#discussion_r1333501889